### PR TITLE
Refactor builders and switch to docker(-py) 2

### PR DESCRIPTION
**Pull request by fbochu**

### CODE REVIEW

File: src/grocker/__main__.py [hunk 1]
```diff
@@ -105,13 +106,18 @@ def __init__(self, option_strings, *args, **kwargs):
         super(PurgeAction, self).__init__(
             option_strings=option_strings,
             dest=argparse.SUPPRESS,
-            choices=('all', 'builders', 'dangling', 'runners'),
+            choices=('old', 'old:runner', 'all', 'all:runner'),
```
Comment (xavfernandez) [inline src/grocker/__main__.py]: This deserves a changelog :o

File: src/grocker/cleanners.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+import docker.errors
+import packaging.version
+
+from . import __version__
+
+logger = logging.getLogger(__name__)
+
+
+def version_compare(obj, version):
```
Comment (xavfernandez) [inline src/grocker/cleanners.py]: This function name is quite unclear.
You could create a function `get_grocker_version` on one end and build a `packaging.specifiers.SpecifierSet` from `__version__`

File: ChangeLog.rst [hunk 3]
```diff
@@ -6,6 +6,8 @@ ChangeLog
 
 - Add support for Python 3.6
 - Drop obsolete .grocker file
+- Switch to docker(-py) 2
+- Rewrite purge option to only mange Grocker created stuff
```
Comment (xavfernandez) [inline ChangeLog.rst]: option**s** and m**a**nage

File: docs/advanced.rst [hunk 4]
```diff
@@ -11,12 +11,12 @@ Command line interface
...
-                   [--image-base-name <name>] [-n <name>]
-                   [--result-file <filename>]
-                   [--purge {all,builders,dangling,runners}] [--version] [-v]
-                   <action> [<action> ...] <release>
+               [--volume VOLUMES] [--port PORTS] [--pip-conf <file>]
+               [--pip-constraint <file>] [--docker-image-prefix <url>]
+               [--image-base-name <name>] [-n <name>]
+               [--result-file <filename>]
+               [--purge {old,old:runner,all,all:runner}] [--version] [-v]
+               <action> [<action> ...] <release>
```
Comment (xavfernandez) [inline docs/advanced.rst]: Wondering if we should be using something like https://github.com/pypa/pip/blob/master/docs/pipext.py#L14-L22
Which automatically creates: https://pip.pypa.io/en/stable/reference/pip_install/#usage (and keep it in sync).

File: src/grocker/__main__.py [hunk 5]
```diff
@@ -105,13 +106,18 @@ def __init__(self, option_strings, *args, **kwargs):
...
-        builders.docker_purge_images(builders.docker_get_client(), filters)
-        builders.docker_purge_volumes(builders.docker_get_client(), filters)
+        current_version = 'all' in values
+        runner = 'runner' in values
+
+        loggers.setup()
+        docker_client = utils.docker_get_client()
+        cleanners.docker_purge_container(docker_client, current_version=current_version)
+        cleanners.docker_purge_volumes(docker_client, current_version=current_version)
+        cleanners.docker_purge_images(docker_client, current_version=current_version, runner=runner)
```
Comment (xavfernandez) [inline src/grocker/__main__.py]: Now that I understand where `PurgeAction` is dealt with, this makes me sad.
`grocker --purge runners --help` should not be doing anything.

File: src/grocker/__main__.py [hunk 6]
```diff
@@ -152,47 +158,40 @@ def main():
...
             config=config,
+            name=image_name,
             release=args.release,
-            wheels_volume_name=wheels_volume_name,
-            tag=image_name,
         )
 
     if GrockerActions.push_img in args.actions:
-        if '/' not in image_name:
+        if builders.is_prefixed_image(image_name):
```
Comment (xavfernandez) [inline src/grocker/__main__.py]: Looks like it is missing a `not` ?

File: docs/advanced.rst [hunk 7]
```diff
@@ -5,79 +5,69 @@ Grocker build chain can be customized using the command line interface or the
...
+      -c, --config <filename>         Grocker config file
+      -r, --runtime <runtime>         runtime used to build and run this image
+      --pip-conf <filename>           pip configuration file used to download
+                                      dependencies (by default use pip config
+                                      getter)
+      --pip-constraint <filename>     pip constraint file used to download
+                                      dependencies
+      -e, --entrypoint TEXT           Docker entrypoint to use to run this image
+      --volume TEXT                   Container storage and configuration area
+      --port TEXT                     Port on which a container will listen for
```
Comment (xavfernandez) [inline docs/advanced.rst]: Strange that we sometimes have `TEXT` and sometimes `<more_explicit_name>`.

File: src/grocker/__main__.py [hunk 8]
```diff
@@ -191,11 +191,12 @@ def main():
...
 
     if GrockerActions.push_img in args.actions:
-        if builders.is_prefixed_image(image_name):
+        if not builders.is_prefixed_image(image_name):
             logger.warning('Not pushing any image since the registry is unclear in %s', image_name)
         else:
             logger.info('Pushing image...')
-            results['hash'] = builders.docker_push_image(docker_client, image_name)
+            image = builders.docker_push_image(docker_client, image_name)
+            collect['hash'] = [x.split('@')[1] for x in image.attrs['RepoDigests']][0]
```
Comment (xavfernandez) [inline src/grocker/__main__.py]: This should still be named results in this commit.
The rename only happens with the CLI switch

File: src/grocker/__main__.py [hunk 9]
```diff
@@ -22,188 +17,134 @@
...
+    help="base name for the image (eg '<image-prefix>/<image-base-name>:<image-version>')",
+)
+@click.option('-n', '--image-name', metavar='<name>', help="name used to tag the build image")
+@click.option(
+    '--result-file', type=click.Path(exists=False), metavar='<filename>',
+    help="yaml file where results (image name, ...) are written",
+)
+@click.option(
+    '--action', multiple=True, type=click.Choice(['all'] + Actions.values()), default=['all'],
+    help='what to do, "all" is equivalent to the three others',
```
Comment (xavfernandez) [inline src/grocker/__main__.py]: It would be nice to also include the three possible values in the help text

File: ChangeLog.rst [hunk 10]
```diff
@@ -6,6 +6,9 @@ ChangeLog
 
 - Add support for Python 3.6
 - Drop obsolete .grocker file
+- Switch to docker(-py) 2
+- Rewrite purge options to only manage Grocker created stuff
+- Switch cli to click (**breaking change**)
```
Comment (xavfernandez) [inline ChangeLog.rst]: Maybe put the breaking change at the top of the changelog

File: ChangeLog.rst [hunk 11]
```diff
@@ -6,6 +6,9 @@ ChangeLog
 
 - Add support for Python 3.6
 - Drop obsolete .grocker file
+- Switch to docker(-py) 2
+- Rewrite purge options to only manage Grocker created stuff
```
Comment (xavfernandez) [inline ChangeLog.rst]: This changelog makes less sens with the CLI complete rewrite

File: docs/advanced.rst (hunk 4 above)
Comment (fbochu) [inline docs/advanced.rst]: I think it's a good idea, can you submit a PR to do that ?

**Pull request merged**